### PR TITLE
refactor: make the control plane the sole reader of raw Hermes frames

### PR DESCRIPTION
**Pull request by jakubswierczek**
## Summary

- The control plane (`src/lib/hermes-control-plane/`) is now the **only** reader of raw Hermes frames on the live path. Classification happens once at gateway ingress; the per-session live buffer stores typed `JuneHermesEvent`s; the transcript builder switches on `kind` and every raw-frame helper it used is deleted.
- **A1 — union widening**: `receivedAt` on every kind; `pending_action_resolution` for the four `*.response` frames (previously `unsupported`); `steering` as a locally minted first-party kind (a steer is never a wire frame — glossary updated); transcript `failed` flag; tool events carry `key`/`text`/`isClarify` and a failed-capable phase over the whole `tool.*` family; subagent task metadata. Divergence policy: where the old duplicate parsers disagreed, transcript-visible behavior wins, and each adopted choice has a rationale comment at the field.
- **A2 — builder migration**: `appendLiveHermesEvents` consumes the union; the previously hand-maintained duplicates (subagent phase rules, the secret-redaction invariant "key name only, never value") now live once, in the classifier. Builder suites construct typed fixtures.
- **B — status split**: terminal vocabulary has one home in the control plane (`isTerminalHermesEvent` + a type-derived lifecycle `flavor`); the activity store imports it instead of mirroring it. The `working`/`waiting`/`toolCall` Sets became an identity-stable projection of store phase; `stopping`/`runtime` stay local intent. Edge timing (unlisten, steer resend, issue promotion) is unchanged, including `message.complete` as per-turn terminal.
- **Review fixes**: a 27-agent review of the branch confirmed 10 findings, all fixed with regression tests — most notably: respond RPCs keep the wire (runtime) session id while only the activity store gets stored-id keying; store eviction prefers inactive rows; a subagent error no longer strands the working spinner; the projection reuses Set identities so streamed tokens don't tear down effects.
- Security-relevant: the secret-redaction invariant is now asserted at a single seam (classifier) instead of duplicated across two files.

## Validation

- `pnpm typecheck`, `pnpm check` (Biome, 0 errors), `pnpm test`: 127 files, **1,935 passed / 0 failed** (includes new classifier, replay-corpus, store, builder, and workspace regression tests).
- Recorded-fixture replay corpus updated: the four `*.response` fixtures now pin `pending_action_resolution` instead of `unsupported`.

- [ ] Tested visually where it matters (UI changes: attach a screenshot or recording). 
- [ ] Needs a June API (backend) deploy before it works end to end. 

## Out of scope

- Stage D of the design: a stateful `TranscriptAccumulator` (incremental fold instead of rebuild-last-200-per-render). Deliberately deferred until this ships and only if perf/ordering bugs surface.
- Unifying the tool dedupe `key` (builder precedence) with `toolCallId` (store precedence) — two precedences kept on purpose for behavior preservation.
- `pending_action_resolution` frames resolving pendingActionStore rows directly (today rows still clear on session terminal, as on main).
- Rust shell and June API untouched.

## Followups

- Consider a small golden raw→classify→build snapshot suite as a divergence backstop now that builder fixtures are typed.
- Migrate the artifact store's remaining payload interpretation onto typed fields when it next changes.

## Security and release checklist

- [x] No secrets, local `.env` values, signing material, tokens, or customer data are committed.
- [x] Security-sensitive changes were called out in the summary.
- [x] Workflow action references are pinned to full-length commit SHAs. 
- [ ] Desktop signing, updater, entitlement, capability, or release changes have owner review. 
- [ ] Backend auth, billing, metering, CORS, rate limit, or logging changes have owner review. 
- [x] User-facing copy follows the repo UI conventions. 

https://claude.ai/code/session_016jdLY1DsSVVAhJKCG8nZmS

### CODE REVIEW

File: src/components/agent/AgentWorkspace.tsx [hunk 1]
```diff
@@ -11834,80 +11674,91 @@ function sessionHasActiveWork(
...
-function agentStatusSummaryFromHermesEvent(
-  event: HermesGatewayEvent,
-  status: AgentSessionStatusKind,
-) {
+function agentStatusSummaryFromHermesEvent(event: JuneHermesEvent, status: AgentSessionStatusKind) {
   if (status === "waitingForUser") {
-    return event.type === "approval.request" ? "June needs approval." : "June has a question.";
+    if (event.kind !== "pending_action") return "June has a question.";
+    return event.action.kind === "clarify" ? "June has a question." : "June needs approval.";
   }
```
[the remark this answers is not in the document]
Comment (jakubswierczek) [reply src/components/agent/AgentWorkspace.tsx:11749]: Fixed in 64c341aa — restored main's exact copy: only approval requests say "June needs approval."; sudo/secret keep "June has a question." (with a comment noting the parity is deliberate).

File: src/lib/hermes-activity-store.ts [hunk 2]
```diff
@@ -308,25 +320,41 @@ function applyEvent(row: InternalRecord, event: JuneHermesEvent): void {
...
+      return;
     case "background_activity":
       applyBackgroundActivity(row, event);
       return;
     case "error":
       row.phase = "error";
       return;
     case "lifecycle":
-      row.phase = isTerminalLifecycleStatus(event.status) ? "complete" : "running";
+      row.phase = isTerminalHermesLifecycleStatus(event.status) ? "complete" : "running";
```
[the remark this answers is not in the document]
Comment (jakubswierczek) [reply src/lib/hermes-activity-store.ts]: Fixed in 64c341aa — the store's lifecycle arm now treats `flavor === "terminal" || isTerminalHermesLifecycleStatus(status)` as terminal, so a terminal frame carrying a non-terminal status word (e.g. turn.complete + status "success") reads complete. The status fallback stays for lifecycle.* frames whose payload status says completed.

File: src/components/agent/AgentWorkspace.tsx [hunk 3]
```diff
@@ -4339,7 +4204,10 @@ export function AgentWorkspace({
       // derives the session's phase (running/waiting/background/error/complete),
       // current tool, and subagent count from the normalized event — never from
       // the raw frame (raw JSON belongs to feature 15's trace panel).
-      hermesActivityStore.record(classified, hermesModeFor(storedSessionId));
+      hermesActivityStore.record(
+        withStoredHermesSessionId(classified, storedSessionId),
+        hermesModeFor(storedSessionId),
```
[the remark this answers is not in the document]
Comment (jakubswierczek) [reply src/components/agent/AgentWorkspace.tsx]: Fixed in 64c341aa — pendingActionStore.record now receives the stored-id-rewritten event at ingress (same as the activity store), and the respond handlers resolve via the stored buffer key, so the drawer's pending-count match holds on resumed runs. The respond RPCs keep the wire session id.

**Comment by jakubswierczek** [comment]
Re: Octopus Review findings (no inline threads to reply on, so consolidating here):

- **Steer-drain trigger widened to any `tool.*complete*` subtype** — deliberate, and annotated at the call site. Failure-flavored subtypes (`tool.error`/`tool.fail`) classify as `failed` and never drain; the only divergence from the old exact `tool.complete` match is a hypothetical non-exact completion subtype, where draining is the correct reading (a tool result was produced).
- **Continuity restore effect reading `workingSessionIdsRef.current`** — the effect is mount-only and the ref is initialized during that same render from the projection of the module-global activity store, which survives workspace unmount and is populated synchronously. At mount time the ref cannot be stale.
- **`clearSessionActivity` read-after-record timing** — `hermesActivityStore.record()` is a synchronous Map mutation plus a version bump; there is no batching or debounce, so the immediate counts read reflects the terminal transition.

**Comment by jakubswierczek** [comment]
## Consolidated review — two-axis (Claude) + adversarial (Codex)

### Two-axis review (vs `main`)

**Standards axis: 0 violations.** Sentence case, dash rules, icon policy, design tokens, and the CONTEXT.md binding `_Avoid_` terminology all check out across the diff. One non-violation note: the new classifier comments are dense but match the module's existing invariant-heavy idiom.

**Spec axis: faithful to the agreed design; 3 flags, dispositions below.**
- *D5 partial:* 4 tests in `agent-chat-runtime.test.ts` (~1041, 1052, 1384) still classify raw frames instead of constructing typed fixtures. Defensible as seam-boundary cases (`message.completed`→lifecycle, error-summary); open cleanup.
- *Store error-guard change:* transcript/reasoning deltas now lift an errored store row back to `running`. Deliberate (review fix for the sticky-error spinner strand); true wire-error terminals unlisten immediately, so no post-error delta can arrive through the listener.
- *Sudo/secret `waitingForUser` dispatch edge (main dispatched only clarify/approval):* forced by the store projection — the store has always mapped all four pending-action kinds to `waiting`, so the status helper matching it keeps dispatch and display consistent.

### Codex adversarial review (verdict: needs-attention, 2 findings)

1. **[high] `lifecycle.complete` classifies as non-terminal** — the workspace terminal path (unlisten, steer resend, pending-action resolution) never runs for it while the store can read it complete from the status string (split-brain).
2. **[medium] Live artifacts recorded under the runtime session id** — the artifact store gets the wire-id event while the drawer reads by stored id, so a resumed run's artifacts can be missed.

**Disposition: both verified as pre-existing on `main`, not branch regressions.** `lifecycle.complete` sits in the raw-type union on main but was never in main's terminal type list, and main's ingress also passed the wire-id event to the artifact store. This branch's binding policy was bit-for-bit behavior preservation, so both gaps were carried over deliberately.

### Follow-ups proposed

- Artifact store: pass the stored-id-rewritten event (one-line alignment with the activity/pending-action stores) + a runtime-vs-stored skew regression test. Cheap; could land in this PR if reviewers prefer.
- `lifecycle.complete` terminal-ness: a real behavior change vs main (affects unlisten/steer timing) — should be decided and shipped separately, not silently inside a parity refactor.
- Convert the 4 raw-classifying tests to typed fixtures, moving their seam assertions into the classifier suite.

https://claude.ai/code/session_016jdLY1DsSVVAhJKCG8nZmS

**Comment by jakubswierczek** [comment]
All three follow-ups from the consolidated review are now implemented in 28d7cbea:

- **`lifecycle.complete`/`.completed` classify as terminal** (deliberate extension beyond main, per the split-brain finding) — the workspace terminal path now runs for them; classifier + workspace-level regression tests added.
- **Artifact store keys by stored session id** — `storedClassified` at ingress like the other stores; runtime-vs-stored skew regression test added.
- **Typed-fixture conversion complete** — no builder-direct suite imports `classifyHermesEvent` anymore; the seam assertions live in the classifier suite.

Gate: typecheck + Biome clean, 1,940 tests passing.
